src/nvim/rbuffer.c

```diff
@@ -34,17 +34,20 @@ void rbuffer_free(RBuffer *buf)
   xfree(buf);
 }
 
-size_t rbuffer_size(RBuffer *buf) FUNC_ATTR_NONNULL_ALL
+size_t rbuffer_size(RBuffer *buf)
+  FUNC_ATTR_NONNULL_ALL FUNC_ATTR_PURE
 {
   return buf->size;
 }
 
-size_t rbuffer_capacity(RBuffer *buf) FUNC_ATTR_NONNULL_ALL
+size_t rbuffer_capacity(RBuffer *buf)
+  FUNC_ATTR_NONNULL_ALL FUNC_ATTR_PURE
 {
   return (size_t)(buf->end_ptr - buf->start_ptr);
 }
 
-size_t rbuffer_space(RBuffer *buf) FUNC_ATTR_NONNULL_ALL
+size_t rbuffer_space(RBuffer *buf)
+  FUNC_ATTR_NONNULL_ALL FUNC_ATTR_PURE
```

Comment:
Are you sure it is safe to mark these functions as pure? Considering `RBuffer` is used by both the main thread and the ui thread, the compiler may not be able to when `RBuffer` is changed by another thread.